### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by mtneug**
This adds a Dockerfile for building Tobira container images. The `x.sh` script was extended with a new `build-container-image` target for doing so as the Dockerfile requires multiple arguments to be passed in.

This partially addresses #624.

### CODE REVIEW

File: util/scripts/build-container-image.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+NODE_VERSION=19.3
+RUST_VERSION=1.66
+
+CONTAINER_PLATFORM="${CONTAINER_PLATFORM:-linux/amd64}"
+RUST_TARGET="${RUST_TARGET:-x86_64-unknown-linux-musl}"
+
+BUILT_TIME_UTC="$(date -u +"%Y-%m-%dT%TZ")"
+GIT_COMMIT_HASH="$(git rev-parse --short HEAD || echo "unknown")"
+VERSION="$(grep -e '^version' backend/Cargo.toml | cut -d '=' -f 2- | tr -d ' "')"
```
Comment (LukasKalbertodt) [inline util/scripts/build-container-image.sh:16]: FWIW this is used in `make-release.sh`:

```sh
sed -n -E 's/^version = "([^"]+)"$/\1/p' backend/Cargo.toml
```

Dunno if either one of those is "better". One could also extract this into `print-version.sh` or sth, but it's also not very important :D

File: util/containers/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+FROM --platform=${BUILDPLATFORM} docker.io/node:${NODE_VERSION}-alpine AS frontend-build
+
+WORKDIR /build
+COPY frontend/package.json frontend/package-lock.json /build/
+RUN npm i
+
+COPY frontend .
+RUN npm ci
+RUN npx relay-compiler
+RUN npx webpack --progress --mode=production
```
Comment (LukasKalbertodt) [inline util/containers/Dockerfile:16]: Is `--progress` well behaved inside `docker build`? Or does it screw up the terminal output? In the latter case, I would just remove it.

File: util/containers/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+FROM --platform=${BUILDPLATFORM} docker.io/node:${NODE_VERSION}-alpine AS frontend-build
+
+WORKDIR /build
+COPY frontend/package.json frontend/package-lock.json /build/
+RUN npm i
+
+COPY frontend .
+RUN npm ci
```
Comment (LukasKalbertodt) [inline util/containers/Dockerfile:14]: I don't understand this part. Why do we first copy the package files and `npm i` when we then copy everything and `npm ci` anyway? To me it seems like the first `COPY` and `RUN npm i` could be removed?

File: util/containers/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  perl
+
+ARG RUST_TARGET
+RUN rustup target add "${RUST_TARGET}"
+
+COPY backend/Cargo.toml backend/Cargo.lock /build/backend/
+COPY <<EOF /build/backend/src/main.rs
+fn main() {}
+EOF
+RUN cargo build --target "${RUST_TARGET}" --release
```
Comment (LukasKalbertodt) [inline util/containers/Dockerfile:38]: Again this: can you explain why this is necessary/useful?

File: util/containers/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      ca-certificates-bundle \
+      musl \
+ && rm -rf \
+      /rootfs/etc/apk \
+      /rootfs/lib/apk \
+      /rootfs/var/cache/apk
+COPY --from=backend-build /build/backend/tobira /rootfs/tobira
+RUN mkdir -p /rootfs/etc/tobira/ \
+ && /rootfs/tobira write-config /rootfs/etc/tobira/config.toml \
+ && sed -i 's/^#address.*/address = "0.0.0.0"/g' /rootfs/etc/tobira/config.toml
```
Comment (LukasKalbertodt) [inline util/containers/Dockerfile:63]: I'm a bit uncomfortable with this as this will affect any other config values called "address" that we might add in the future. And we would likely not catch it if we don't happen to think of this Dockerfile. I don't really have a better idea right now, but mhhh

File: util/containers/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,86 @@
+# syntax=docker/dockerfile:1.4
```
Comment (LukasKalbertodt) [inline util/containers/Dockerfile:1]: General remark: this file duplicates quite a bit of the logic inside `build-release.sh`. Would it be possible to avoid the duplication? In particular the annoying `prefix-map` thingies below are things that one probably doesn't want to repeat. 

If we ever need to change anything about the build process, and even if we don't forget about this file, it's tricky to adjust and test this. So also from a maintenance perspective, I would prefer the Dockerfile to use `build-release.sh`. 

While I don't know a lot about Docker, I realize that there might be good reasons to do it this way. Each step here adds a new fs layer or something like that? But all the files that are created during the build (except the binary) are not part of the final container right? So it shouldn't matter? 
So yeah, what do you think?

File: .dockerignore [hunk 7]
```diff
@@ -0,0 +1,14 @@
+.deployment/
```
Comment (LukasKalbertodt) [inline .dockerignore:1]: Is this file really necessary? What docker command would automatically do what with all the files mentioned here otherwise? Seems like the Dockerfile already copies only very specific files and folders?

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Thanks for the PR! Just to make sure I understand the goal of this correctly: this adds all the scripts/definitions for someone to easily upload an image to `quay.io`? And that image could be used by others to quickly setup Tobira? How would the user of this image change the `config.toml`?

And there would be the question of maintenance. Who can upload to quay.io? Would you maintain this, i.e. send PRs whenever the Dockerfile needs adjustments?

File: util/containers/Dockerfile (hunk 3 above)
Comment (mtneug) [reply util/containers/Dockerfile:14]: In terms of a correct build, yes. This pattern is common to make efficient use of Docker build caches. Every instruction will create a new layer in the current image (note that there can be multiple stages in a build, always starting with a `FROM`). 

By first running `npm i`, this layer will already include all downloaded npm packages. If `packages.json` and `package-lock.json` do not change between builds, Docker will simply use the already built layer containing the packages. Builds will effectively start from the `COPY frontend .` command, and since all packages are already downloaded, `npm ci` will be much faster as well.

File: util/containers/Dockerfile (hunk 4 above)
Comment (mtneug) [reply util/containers/Dockerfile:38]: Same principle as with npm. This will cache a build of all Rust dependencies. If versions do not change, these will not be rebuilt again.

There is a relatively new alternative that could be used: instead of caching in layers, one can mount a caching directory during the build where output between builds can be reused. Would you prefer that?

File: util/containers/Dockerfile (hunk 5 above)
Comment (mtneug) [reply util/containers/Dockerfile:63]: Thought the same, with no better alternative than to create a static config file for the container image. But that needs updating.

File: util/containers/Dockerfile (hunk 6 above)
Comment (mtneug) [reply util/containers/Dockerfile:1]: Using the alternative approach by mounting in the cached build output, it might be possible to reuse `build-release.sh`. But the build is relatively straightforward. Yes, the final image is very minimal, only containing the binary + some config files/libs. There should be no shell etc. Effectively: a Linux distribution containing the Alpine packages `alpine-baselayout-data`, `ca-certificates-bundle`, and `musl` + Tobira.

File: .dockerignore (hunk 7 above)
Comment (mtneug) [reply .dockerignore:1]: Yes, not every file is used in the `Dockerfile`. I'm pretty sure, they will still be uploaded to the Docker daemon if not ignored. Running a container build will upload two files: the Dockerfile + the build context. The build context is a folder (here the root project folder) excluding every file from `.dockerignore`.

**Comment by mtneug** [comment]
> Thanks for the PR! Just to make sure I understand the goal of this correctly: this adds all the scripts/definitions for someone to easily upload an image to `quay.io`?

Well first of all build, but since the image name is prefixed with `quay.io`, it can be pushed directly.

> And that image could be used by others to quickly setup Tobira? How would the user of this image change the `config.toml`?

Yes.

I would suggest to mount a `config.toml` into the container. Directly using Docker: `docker run .... -v /host/config.toml:/etc/tobira/config.toml:ro ....`. In a Kubernetes cluster, one would create a `Secret` and mount that in.

> And there would be the question of maintenance. Who can upload to quay.io? Would you maintain this, i.e., send PRs whenever the Dockerfile needs adjustments?

I can upload to quay.io and I think @lkiesow can as well. Ideally, this would be part of a GitHub action workflow. BTW for other projects, I started using the GitHub container registry. That could be easier to integrate, but both shouldn't be hard. I can maintain the Dockerfile if build instructions change. Though I will likely not follow things too closely. A heads-up would be appreciated ;)

File: util/containers/Dockerfile (hunk 5 above)
Comment (LukasKalbertodt) [reply util/containers/Dockerfile:63]: This seems to work (i.e. only modifies `http.address`).

```suggestion
 && sed -i '/\[http\]/,/^#address =.*$/ s/^#address =.*$/address = "0.0.0.0"/' /rootfs/etc/tobira/config.toml
```

**Comment by LukasKalbertodt** [comment]
> I can upload to quay.io and I think @lkiesow can as well. Ideally, this would be part of a GitHub action workflow. BTW for other projects, I started using the GitHub container registry. That could be easier to integrate, but both shouldn't be hard. I can maintain the Dockerfile if build instructions change. Though I will likely not follow things too closely. A heads-up would be appreciated ;)

Sounds good! Then we can merge it. Automation can still be added later.

---

Regarding the inline comments about caching and using `build-release.sh`: I don't have the experience to decide anything here. Using this alternative caching thing and then using `build-release.sh` sounds good to *me*, but you decide. In the end, you are basically maintaining it :P

File: util/containers/Dockerfile (hunk 3 above)
Comment (LukasKalbertodt) [reply util/containers/Dockerfile:14]: Thanks for the explanation. See main-thread comment below for more comments.

File: util/containers/Dockerfile (hunk 4 above)
Comment (LukasKalbertodt) [reply util/containers/Dockerfile:38]: See main-thread comment below for more comments.

File: util/containers/Dockerfile (hunk 6 above)
Comment (LukasKalbertodt) [reply util/containers/Dockerfile:1]: As I wrote below: it sounds like a good idea to me to try out this alternative caching thingy and then using `build-release.sh`, but it's your call.

File: backend/src/version.rs [hunk 8]
```diff
@@ -17,7 +17,8 @@ pub(crate) fn build_time_utc() -> &'static str {
 
 /// Returns the commit hash this was build from.
 pub(crate) fn git_commit_hash() -> &'static str {
-    build_info::GIT_COMMIT_HASH.expect("missing git version info")
+    return build_info::GIT_COMMIT_HASH.unwrap_or_else(
+        || option_env!("GIT_COMMIT_HASH").expect("missing git version info"))
```
[the remark this answers is not in the document]
Comment (LukasKalbertodt) [reply backend/src/version.rs]: Woops, i have one `)` too much, which makes CI fail. Sorry for that, never trust my suggestions :sweat_smile:

**Comment by mtneug** [comment]
> Regarding the inline comments about caching and using build-release.sh: I don't have the experience to decide anything here. Using this alternative caching thing and then using build-release.sh sounds good to me, but you decide. In the end, you are basically maintaining it :P

I would leave it as is, as this is still more common.

**Comment by LukasKalbertodt** [comment]
I should really test my code suggestions before posting them in code reviews :P To not bother you again I simply fixed it. I also adjusted the `README` in the containers folder slightly. With that, this is good to go!
